src/zjs_dgram.c

```diff
@@ -0,0 +1,334 @@
+                                            const jerry_value_t this,
+                                            const jerry_value_t argv[],
+                                            const jerry_length_t argc)
+{
+    int ret;
+    if (argc != 1)
+        return zjs_error("createSocket: invalid argument");
+
+    char type_str[8];
+    GET_STR(argv[0], type_str);
+
+    sa_family_t family;
+    if (strcmp(type_str, "udp4") == 0)
+        family = AF_INET;
+    else if (strcmp(type_str, "udp6") == 0)
+        family = AF_INET6;
+    else
+        return zjs_error("createSocket: invalid argument");
+
+    struct net_context *udp_sock;
+    CHECK(net_context_get(family, SOCK_DGRAM, IPPROTO_UDP, &udp_sock));
+
+    jerry_value_t sockobj = jerry_create_object();
+    jerry_set_prototype(sockobj, zjs_dgram_socket_prototype);
+
+    dgram_handle_t *handle = zjs_malloc(sizeof(handle));
+    if (!handle)
+        return zjs_error("createSocket: OOM");
+    handle->udp_sock = udp_sock;
+    handle->message_cb_id = -1;
+
+    jerry_set_object_native_handle(sockobj, (uintptr_t)handle, zjs_dgram_free_cb);
+
+    // Can't call this here due to bug in Zephyr - called in .bind() instead
+    //CHECK(net_context_recv(udp_sock, udp_received, K_NO_WAIT, handle));
+
+    return sockobj;
+}
+
+static jerry_value_t zjs_dgram_sock_on(const jerry_value_t function_obj,
+                                    const jerry_value_t this,
+                                    const jerry_value_t argv[],
+                                    const jerry_length_t argc)
+{
+    if (argc < 2 ||
+        !jerry_value_is_string(argv[0]) ||
+        (!jerry_value_is_function(argv[1]) &&
+         !jerry_value_is_null(argv[1]))) {
+        return zjs_error("zjs_dgram_sock_on: invalid argument");
+    }
+
+    GET_HANDLE(dgram_handle_t, handle);
+
+    jerry_size_t str_sz = 32;
+    char event[str_sz];
+    zjs_copy_jstring(argv[0], event, &str_sz);
+    if (strcmp(event, "message"))
+        return zjs_error("zjs_dgram_sock_on: unsupported event type");
+
+    if (handle->message_cb_id != -1)
+        zjs_remove_callback(handle->message_cb_id);
+    if (!jerry_value_is_null(argv[1]))
+        handle->message_cb_id = zjs_add_callback(argv[1], this, handle, NULL);
+
+    return ZJS_UNDEFINED;
+}
+
+// Zephyr "packet sent" callback
+static void udp_sent(struct net_context *context, int status, void *token,
+                     void *user_data)
+{
+    DBG_PRINT("udp_sent: %p, st=%d udata=%p\n", context, status, user_data);
+
+    if (user_data) {
+        jerry_value_t args[1] = {ZJS_UNDEFINED};
+        if (status != 0) {
+            char errbuf[8];
+            snprintf(errbuf, sizeof(errbuf), "%d", status);
+            args[0] = zjs_standard_error(NetworkError, errbuf);
+            // We need error object, not error value (JrS doesn't allow to
+            // pass the latter as a func argument).
+            jerry_value_clear_error_flag(&args[0]);
```

Comment:
Interesting, with this comment and this line I think you understand something important that I don't understand. :) Please chat with me about this in IRC at some point.